### PR TITLE
Refactor PagedDoraWorker by injecting MetaManager and UfsManager

### PR DESCRIPTION
**Pull request by jiacheliu3**
### What changes are proposed in this pull request?

1. If an object is created inside `PagedDoraWorker` constructor, extract that creation to before the constructor and use dependency injection to inject it to the worker object. This doesn't change any creation logic, just a refactor to better adapt to dependency injection flavor.
2. There is a circular dependency between `MetaManager` and `PagedDoraWorker`. This change removes that cycle. Now we create one, then create the other. Before, we create one and in the construction, we let `this` ref escape and create the other. Some methods are either moved or changed to `static`.
3. By adapting to dependency injection, we rely on `UfsManager` interface instead of `DoraUfsManager` implementation. Some method signatures are extracted to the interface level.
4. A few other small refactors to get rid of some downcasts and variable scope changes. Reasons are attached in comments on this PR.

### Why are the changes needed?

Improve code quality and extensibility.

### Does this PR introduce any user facing changes?

No. All refactor changes are small and equivalent to existing code. So nothing should break.

### CODE REVIEW

File: dora/core/server/worker/src/test/java/alluxio/worker/dora/DoraMetaManagerTest.java [hunk 1]
```diff
@@ -57,7 +57,7 @@ public void before() throws IOException {
     PagedDoraWorker worker = mock(PagedDoraWorker.class);
     CacheManager cacheManager = mock(CacheManager.class);
     mDoraUfsManager = mock(DoraUfsManager.class);
-    mManager = new DoraMetaManager(conf, worker, cacheManager, mDoraUfsManager);
+    mManager = new DoraMetaManager(conf, cacheManager, mDoraUfsManager);
```
Comment (jiacheliu3) [inline dora/core/server/worker/src/test/java/alluxio/worker/dora/DoraMetaManagerTest.java:60]: this line is one of the core change in this PR. The PagedDoraWorker no longer passes its `this` out in its own constructor. The `this` ref escaping in the constructure is a bad practice we should avoid.

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 2]
```diff
@@ -184,23 +184,15 @@ public PagedDoraWorker(
...
-      FileSystemContext fileSystemContext) {
+      DoraUfsManager ufsManager,
+      DoraMetaManager metaManager,
+      FileSystemContext fileSystemContext
+  ) {
     super(ExecutorServiceFactories.fixedThreadPool("dora-worker-executor", 5));
     mWorkerId = workerId;
     mConf = conf;
-    mUfsManager = mResourceCloser.register(new DoraUfsManager());
+    mUfsManager = mResourceCloser.register(ufsManager);
```
Comment (jiacheliu3) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:195]: I'm just creating the UfsManager outside this constructor and passing it in. The code logic does not change at all so this refactor should be effectively minor.

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 3]
```diff
@@ -209,8 +201,7 @@ protected PagedDoraWorker(
     mPageSize = mConf.getBytes(PropertyKey.WORKER_PAGE_STORE_PAGE_SIZE);
     mBlockMasterClientPool = blockMasterClientPool;
     mCacheManager = cacheManager;
-    mMetaManager = mResourceCloser.register(
-        new DoraMetaManager(mConf, this, mCacheManager, mUfsManager));
+    mMetaManager = mResourceCloser.register(metaManager);
```
Comment (jiacheliu3) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:204]: Interestingly, I don't have something like `bind(MetaManager).to(DoraMetaManager.class)` in `DoraWorkerModule` and Guice is actually able to notice that and create the `MetaManager` instance for me. Is that okay? I figure we should always call `bind()` explicitly in code?

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 4]
```diff
@@ -454,83 +445,6 @@ protected int getCachedPercentage(alluxio.grpc.FileInfo fi, String ufsFullPath)
...
 
-  /**
-   * Build FileInfo from UfsStatus and UFS full Path.
-   *
-   * @param status
-   * @param ufsFullPath
-   * @param xattrMap
-   * @return a FileInfo
-   */
-  public alluxio.grpc.FileInfo buildFileInfoFromUfsStatus(UfsStatus status, String ufsFullPath,
```
Comment (jiacheliu3) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:465]: These two methods are just changed `static` and moved to the end of this file.

File: dora/core/server/worker/src/main/java/alluxio/worker/modules/DoraWorkerModule.java [hunk 5]
```diff
@@ -75,6 +75,10 @@ protected void configure() {
     bind(UfsManager.class).to(DoraUfsManager.class).in(Scopes.SINGLETON);
     bind(AlluxioConfiguration.class).toProvider(() -> Configuration.global());
 
+    // Create FileSystemContext shared across all worker components
+    FileSystemContext fileSystemContext = FileSystemContext.create();
+    bind(FileSystemContext.class).toInstance(fileSystemContext);
```
Comment (jiacheliu3) [inline dora/core/server/worker/src/main/java/alluxio/worker/modules/DoraWorkerModule.java:80]: Similar to the change in PagedDoraWorker constructor, I'm just moving some resource creation out of `PagedDoraWorker()` into this module definition, and injecting into the constructor. The code logic effectively doesn't change.

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 6]
```diff
@@ -145,18 +145,18 @@ public class PagedDoraWorker extends AbstractWorker implements DoraWorker {
   // for now Dora Worker does not support Alluxio <-> UFS mapping,
   // and assumes all UFS paths belong to the same UFS.
   private static final int MOUNT_POINT = 1;
-  private final Closer mResourceCloser = Closer.create();
+  protected final Closer mResourceCloser = Closer.create();
```
Comment (jiacheliu3) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:148]: are these changes necessary?

File: dora/core/server/worker/src/main/java/alluxio/worker/DataServerFactory.java [hunk 7]
```diff
@@ -59,9 +60,9 @@ protected DataServerFactory(UfsManager ufsManager,
    */
   public DataServer createRemoteGrpcDataServer(DataWorker dataWorker) {
     BlockWorkerGrpc.BlockWorkerImplBase blockWorkerService;
-    if (dataWorker instanceof DoraWorker) {
+    if (dataWorker instanceof PagedDoraWorker) {
       blockWorkerService =
-          new DoraWorkerClientServiceHandler((DoraWorker) dataWorker);
+          new DoraWorkerClientServiceHandler((PagedDoraWorker) dataWorker);
```
Comment (jiacheliu3) [inline dora/core/server/worker/src/main/java/alluxio/worker/DataServerFactory.java:65]: `DoraWorker` interface is missing a lot of method definitions so we need to downcast in `DoraWorkerClientServiceHandler` after all. So I'm just casting to the real class to avoid the frequent casting.

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 8]
```diff
@@ -149,18 +150,18 @@ public class PagedDoraWorker extends AbstractWorker implements DoraWorker {
...
   // TODO(lucy) change to string typed once membership manager got enabled by default
   private final AtomicReference<WorkerIdentity> mWorkerId;
   protected final CacheManager mCacheManager;
-  protected final DoraUfsManager mUfsManager;
-  private final DoraMetaManager mMetaManager;
+  protected final UfsManager mUfsManager;
```
Comment (jiacheliu3) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:157]: Use interface here since we are using dependency injection. If one method is not defined on the interface (in the concrete class), I'm moving the signature to the interface level.

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 9]
```diff
@@ -149,18 +150,18 @@ public class PagedDoraWorker extends AbstractWorker implements DoraWorker {
...
+  protected DoraMetaManager mMetaManager;
```
Comment (jiacheliu3) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:158]: it's necessary to change the scope here so a child class can see this field

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 10]
```diff
@@ -1219,4 +1139,85 @@ protected DoraOpenFileHandleContainer getOpenFileHandleContainer() {
...
+   * Build FileInfo from UfsStatus and UFS full Path.
+   *
+   * @param cacheUsage
+   * @param ufsType
+   * @param status
+   * @param ufsFullPath
+   * @param xattrMap
+   * @return a FileInfo
+   */
+  public static alluxio.grpc.FileInfo buildFileInfoFromUfsStatus(
```
Comment (jiacheliu3) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:1153]: there's probably a better place for these 2 static methods but I don't see it right now

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java (hunk 2 above)
Comment (dbw9580) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:200]: Guice advises against injecting closeable resources especially when it's a singleton, as it's hard to reason about who is responsible to close the resource. https://github.com/google/guice/wiki/Avoid-Injecting-Closable-Resources

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java (hunk 3 above)
Comment (dbw9580) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:209]: `DoraMetaManager` is not an interface but a concrete class. There is no `MetaManager`.

**Review by dbw9580 — APPROVED** [review]
LGTM

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Comment by jiacheliu3** [comment]
alluxio-bot, cherry-pick this to branch-3-1.1.x please
